### PR TITLE
Add hepmcmerger

### PR DESCRIPTION
**Pull request by kkauder**
### Briefly, what does this PR introduce?
Add a package for https://github.com/eic/HEPMC_Merger

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [ ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No

### Does this PR change default behavior?
No

### CODE REVIEW

File: packages/hepmcmerger/package.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+# ----------------------------------------------------------------------------
+# If you submit this package back to Spack as a pull request,
+# please first remove this boilerplate and all FIXME comments.
```
Comment (wdconinc) [inline packages/hepmcmerger/package.py]: Note this comment: remove the boilerplate and all fixme comments.

**Review by wdconinc — CHANGES_REQUESTED** [review]
Did you test that this actually works? Maybe you didn't submit all changes to this PR since you define cmake_args without defining this package as a CMakePackage.

**Comment by kkauder** [comment]
I couldn't test it, can you tell me how? This is something assembed from looking at existing packages but I'm flying blind

File: packages/hepmcmerger/package.py (hunk 1 above)
Comment (kkauder) [reply packages/hepmcmerger/package.py]: Done

**Comment by wdconinc** [comment]
> I couldn't test it, can you tell me how? This is something assembed from looking at existing packages but I'm flying blind

You can do something like:
```console
docker run --rm -it spack/ubuntu-noble:0.21.0
git clone https://github.com/eic/eic-spack
git -C eic-spack checkout <branch>
spack repo add eic-spack
spack install <pkg>
```
(This could have typos, on my phone.)

**Comment by kkauder** [comment]
After chewing on this command for three hours, it just failed with the unrelated
```
 >> 9686    FAILED: tools/clang/lib/ASTMatchers/Dynamic/CMakeFiles/obj.clangDynamicASTMatchers.dir/Re
             gistry.cpp.o
     9687    /opt/spack/lib/spack/env/gcc/g++ -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_M
```
Is there anything else I can do?

**Comment by wdconinc** [comment]
> Is there anything else I can do?

Not in this PR, apparently. Can you copy the full error message (with more than 'FAILED') and file it on spack?

**Comment by wdconinc** [comment]
> > Is there anything else I can do?
> 
> Not in this PR, apparently. Can you copy the full error message (with more than 'FAILED') and file it on spack?

And tag me in the issue on spack.

File: packages/hepmcmerger/package.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+from spack.package import *
+
+
+class Hepmcmerger(Package):
+    """An EIC HepMC merger to combine signal and background events."""
+
+    # FIXME: Add a proper url for your package's homepage here.
+    homepage = "https://github.com/eic/HEPMC_Merger"
+    url = "https://github.com/eic/HEPMC_Merger/releases/tag/v1.0.4"
```
Comment (wdconinc) [inline packages/hepmcmerger/package.py]: This needs to be a tar.gz file. Your downloading a github webpage.

File: packages/hepmcmerger/package.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    list_url = "https://github.com/eic/HEPMC_Merger/tags"
+    git = "https://github.com/eic/HEPMC_Merger"
+
+    maintainers = ["kkauder"]
```
Comment (wdconinc) [inline packages/hepmcmerger/package.py]: Use directive, not assignment.

File: packages/hepmcmerger/package.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    version("v1.0.1", sha256="f244ff4de311c164597eaedf547a7305b6d3cc1e6b2bc6e6d9acb8e603a44a4b")
+    version("v1.0.0", sha256="c6088ba6ce73a92d165bc82e19f6670a1545fcfb6f352cdf5629c395a95c16af")
+
+    depends_on("hepmc3")
+    depends_on("root")
+
+    def cmake_args(self):
+        args = []
+        args.append(
+            "-DCMAKE_CXX_STANDARD=%s" % self.spec["root"].variants["cxxstd"].value
```
Comment (wdconinc) [inline packages/hepmcmerger/package.py]: Use f-string.

File: packages/hepmcmerger/package.py [hunk 5]
```diff
@@ -29,6 +29,7 @@ class Hepmcmerger(Package):
 
     depends_on("hepmc3")
     depends_on("root")
+    depends_on("cmake", type="build")
```
Comment (wdconinc) [inline packages/hepmcmerger/package.py]: CMakePackage includes cmake as a dependency.
```suggestion
```

File: packages/hepmcmerger/package.py (hunk 2 above)
Comment (kkauder) [reply packages/hepmcmerger/package.py]: Thanks for the catch

File: packages/hepmcmerger/package.py (hunk 3 above)
Comment (kkauder) [reply packages/hepmcmerger/package.py]: I don't know what that comment means

File: packages/hepmcmerger/package.py (hunk 4 above)
Comment (kkauder) [reply packages/hepmcmerger/package.py]: ok

File: packages/hepmcmerger/package.py (hunk 3 above)
Comment (wdconinc) [reply packages/hepmcmerger/package.py]: https://spack.readthedocs.io/en/latest/spack.html#spack.directives.maintainers

```suggestion
    maintainers("kkauder")
```

**Comment by kkauder** [comment]
I undid the f-string. It didn't work and this syntax is literally the same throughout the repo according to `grep -r DCMAKE_CXX_STANDARD`

**Comment by wdconinc** [comment]
Use single quotes inside the f-string arguments.

**Comment by wdconinc** [comment]
See last two examples in https://peps.python.org/pep-0498/#escape-sequences for correct f-string syntax.

File: packages/hepmcmerger/package.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    version("v1.0.1", sha256="f244ff4de311c164597eaedf547a7305b6d3cc1e6b2bc6e6d9acb8e603a44a4b")
+    version("v1.0.0", sha256="c6088ba6ce73a92d165bc82e19f6670a1545fcfb6f352cdf5629c395a95c16af")
+
+    depends_on("hepmc3")
+    depends_on("root")
+
+    def cmake_args(self):
+        args = []
+        args.append(
+            f"-DCMAKE_CXX_STANDARD={self.spec['"root"'].variants['"cxxstd"'].value}"
```
Comment (wdconinc) [inline packages/hepmcmerger/package.py]:
```suggestion
            f"-DCMAKE_CXX_STANDARD={self.spec['root'].variants['cxxstd'].value}"
```

Comment (wdconinc) [reply packages/hepmcmerger/package.py]: or even better
```suggestion
            self.define("CMAKE_CXX_STANDARD", self.spec["root"].variants["cxxstd"].value)
```

**Comment by kkauder** [comment]
Thanks for all the help, @wdconinc ! Is the fact that the checks pass good enough to try including it in eic-shell?

**Comment by wdconinc** [comment]
Not at all. The checks don't do anything ;-)

**Review by wdconinc — APPROVED** [review]
Works
```
$ SignalBackgroundMerger 
==================================================================
=== EPIC HEPMC MERGER ===
authors: Benjamen Sterwerf* (bsterwerf@berkeley.edu), Kolja Kauder** (kkauder@bnl.gov), Reynier Cruz-Torres***
* University of California, Berkeley
** Brookhaven National Laboratory
*** formerly Lawrence Berkeley National Laboratory

For more information, run 
./signal_background_merger --help
Number of Slices:10000
Signal events file and frequency:
        - root://dtn-eic.jlab.org//work/eic2/EPIC/EVGEN/SIDIS/pythia6-eic/1.0.0/10x100/q2_0to1/pythia_ep_noradcor_10x100_q2_0.000000001_1.0_run1.ab.hepmc3.tree.root    (one event per time slice)

Background files and their respective frequencies:
        - root://dtn-eic.jlab.org//work/eic2/EPIC/EVGEN/BACKGROUNDS/BEAMGAS/proton/pythia8.306-1.0/100GeV/pythia8.306-1.0_ProtonBeamGas_100GeV_run082.hepmc3.tree.root  31.900000 kHz
        - root://dtn-eic.jlab.org//work/eic2/EPIC/EVGEN/BACKGROUNDS/BEAMGAS/electron/beam_gas_ep_10GeV_foam_emin10keV_30Mevt.hepmc3.tree.root   3177.250000 kHz
Writing to bgmerged.hepmc3.tree.root
Prepping root://dtn-eic.jlab.org//work/eic2/EPIC/EVGEN/SIDIS/pythia6-eic/1.0.0/10x100/q2_0to1/pythia_ep_noradcor_10x100_q2_0.000000001_1.0_run1.ab.hepmc3.tree.root
Prepping root://dtn-eic.jlab.org//work/eic2/EPIC/EVGEN/BACKGROUNDS/BEAMGAS/proton/pythia8.306-1.0/100GeV/pythia8.306-1.0_ProtonBeamGas_100GeV_run082.hepmc3.tree.root
Prepping root://dtn-eic.jlab.org//work/eic2/EPIC/EVGEN/BACKGROUNDS/BEAMGAS/electron/beam_gas_ep_10GeV_foam_emin10keV_30Mevt.hepmc3.tree.root
Initiation time: 8 sec

==================================================================

Working on slice 1
Current memory usage: 366 MB
Working on slice 101
Current memory usage: 399 MB
Working on slice 201
Current memory usage: 456 MB
<snip>
Current memory usage: 486 MB
Working on slice 9801
Current memory usage: 486 MB
Working on slice 9901
Current memory usage: 486 MB
Finished all requested slices.
Slice loop time: 0 min
 -- 1919 us / slice
From root://dtn-eic.jlab.org//work/eic2/EPIC/EVGEN/BACKGROUNDS/BEAMGAS/electron/beam_gas_ep_10GeV_foam_emin10keV_30Mevt.hepmc3.tree.root
  placed 63568 events
  --> on average 6
  placed 127136 final state particles
  --> on average 12
From root://dtn-eic.jlab.org//work/eic2/EPIC/EVGEN/BACKGROUNDS/BEAMGAS/proton/pythia8.306-1.0/100GeV/pythia8.306-1.0_ProtonBeamGas_100GeV_run082.hepmc3.tree.root
  placed 655 events
  --> on average 0
  placed 8816 final state particles
  --> on average 0
From root://dtn-eic.jlab.org//work/eic2/EPIC/EVGEN/SIDIS/pythia6-eic/1.0.0/10x100/q2_0to1/pythia_ep_noradcor_10x100_q2_0.000000001_1.0_run1.ab.hepmc3.tree.root
  placed 10000 events
  --> on average 1
  placed 133302 final state particles
  --> on average 13

Maximum Resident Memory 486.938 MB

==================================================================
Overall running time: 0 min
```

**Pull request merged**